compose.py

```diff
@@ -0,0 +1,350 @@
+
+
+def end_gpu_dockerfile(ddir, dockerfile_name, argmap):
+    # Install additional dependencies
+    df = ""
+    if argmap['SAGEMAKER_ENDPOINT']:
+        df += '''
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+COPY --chown=1000:1000 --from=full /usr/bin/serve /usr/bin/.
+'''
+    with open(os.path.join(ddir, dockerfile_name), "a") as dfile:
+        dfile.write(df)
+
+
+def build_docker_image(ddir, dockerfile_name, container_name):
+    # Create container with docker build
+    p = subprocess.Popen(['docker', 'build', '-t', container_name, '-f', \
+        os.path.join(ddir, dockerfile_name), '.'])
+    p.wait()
+    fail_if(p.returncode != 0, 'docker build {} failed'.format(container_name))
+
+
+def get_container_version_if_not_specified():
+    if FLAGS.container_version is None:
+        # Read from TRITON_VERSION file in server repo to determine version
+        with open('TRITON_VERSION', "r") as vfile:
+            version = vfile.readline().strip()
+        import build
+        FLAGS.container_version, upstream_container_version = build.get_container_versions(
+            version, FLAGS.container_version, "")
+        log('version {}'.format(version))
+    log('using container version {}'.format(FLAGS.container_version))
+
+
+def create_argmap(images):
+    # Extract information from upstream build and create map other functions can
+    # use
+    upstreamDockerImage = images["full"]
+
+    # first pull docker image
+    log("pulling container:{}".format(upstreamDockerImage))
+    p = subprocess.run(['docker', 'pull', upstreamDockerImage])
+    fail_if(
+        p.returncode != 0,
+        'docker pull container {} failed, {}'.format(upstreamDockerImage,
+                                                     p.stderr))
+
+    baseRunArgs = ['docker', 'inspect', '-f']
+    p_version = subprocess.run(baseRunArgs + [
+        '{{range $index, $value := .Config.Env}}{{$value}} {{end}}',
+        upstreamDockerImage
+    ],
+                               capture_output=True,
+                               text=True)
+    vars = p_version.stdout
+    log_verbose("inspect args: {}".format(vars))
+    import re  # parse all PATH enviroment variables
+    e = re.search("TRITON_SERVER_VERSION=([\S]{6,}) ", vars)
+    version = "" if e == None else e.group(1)
+    fail_if(
+        p_version.returncode != 0 or len(version) == 0,
+        'docker inspect to find triton version failed, {}'.format(
+            p_version.stderr))
+
+    vars = p_version.stdout
+    e = re.search("NVIDIA_TRITON_SERVER_VERSION=([\S]{5,}) ", vars)
+    container_version = "" if e == None else e.group(1)
+    fail_if(
+        len(container_version) == 0,
+        'docker inspect to find triton container version failed, {}'.format(
+            vars))
+
+    vars = p_version.stdout
+    dcgm_ver = re.search("DCGM_VERSION=([\S]{4,}) ", vars)
+    dcgm_version = ""
+    if dcgm_ver == None:
+        dcgm_version = "2.2.3"
```

Comment:
Why was this changed back to 2.2.3?

Reply:
Since the two versions that matter -- 21.06 and 21.07 will have 2.2.3, we should just use that for the older versions. 21.08 will have 2.2.8,

Reply:
Note the buildbase is the only time we will need the patch anyway, so this should work (and additionally make the composed container have the same version of dcgm as 21.06 and 21.07